proxy/src/cache/project_info.rs

```diff
@@ -0,0 +1,530 @@
+                    .insert(role_name.clone(), secret.into());
+            }
+        } else {
+            self.cache.insert(
+                endpoint_id.clone(),
+                EndpointInfo::new_with_secret(role_name.clone(), secret),
+            );
+        }
+    }
+    pub fn insert_allowed_ips(
+        &self,
+        project_id: &SmolStr,
+        endpoint_id: &SmolStr,
+        allowed_ips: Arc<Vec<SmolStr>>,
+    ) {
+        if self.cache.len() >= self.config.size {
+            // If there are too many entries, wait until the next gc cycle.
+            return;
+        }
+        self.inser_project2endpoint(project_id, endpoint_id);
+        if let Some(mut endpoint_info) = self.cache.get_mut(endpoint_id) {
+            endpoint_info.allowed_ips = Some(allowed_ips.into());
+        } else {
+            self.cache.insert(
+                endpoint_id.clone(),
+                EndpointInfo::new_with_allowed_ips(allowed_ips),
+            );
+        }
+    }
+    fn inser_project2endpoint(&self, project_id: &SmolStr, endpoint_id: &SmolStr) {
+        if let Some(mut endpoints) = self.project2ep.get_mut(project_id) {
+            endpoints.insert(endpoint_id.clone());
+        } else {
+            self.project2ep
+                .insert(project_id.clone(), HashSet::from([endpoint_id.clone()]));
+        }
+    }
+    fn get_cache_times(&self) -> (Instant, Option<Instant>) {
+        let mut valid_since = Instant::now() - self.config.ttl;
+        // Only ignore cache if ttl is disabled.
+        let ignore_cache_since = if !self.ttl_enabled.load(std::sync::atomic::Ordering::Relaxed) {
+            let ignore_cache_since = *self.ttl_disabled_since.read();
+            // We are fine if entry is not older than ttl or was added before we are getting notifications.
+            valid_since = valid_since.min(ignore_cache_since);
+            Some(ignore_cache_since)
+        } else {
+            None
+        };
+        (valid_since, ignore_cache_since)
+    }
+
+    pub async fn gc_worker(&self) -> anyhow::Result<Infallible> {
+        let mut interval =
+            tokio::time::interval(self.config.gc_interval / (self.cache.shards().len()) as u32);
+        loop {
+            interval.tick().await;
+            self.gc();
+        }
+    }
+
+    fn gc(&self) {
+        if self.cache.len() >= self.config.size {
+            // If there are too many entries, wait until the next gc cycle.
+            return;
+        }
+        let shard = thread_rng().gen_range(0..self.project2ep.shards().len());
+        debug!(shard, "project_info_cache: performing epoch reclamation");
+
+        // acquire a random shard lock
+        let mut removed = 0;
+        let shard = self.project2ep.shards()[shard].write();
+        for (_, endpoints) in shard.iter() {
+            for endpoint in endpoints.get().iter() {
+                self.cache.remove(endpoint);
+                removed += 1;
+            }
+        }
```

Comment:
I'm unsure about this GC loop. It doesn't seem to remove "garbage", instead just random endpoints. 

Because it clears all endpoints in a project2ep shard, that means that every gc_interval, we would expect that all endpoints will be cleared. No cache will survive the gc_interval duration.

I guess for now we can set the values quite high. eg
size = 10000
gc_interval = 60m